File: lib/services/sync_service.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:googleapis/drive/v3.dart' as drive;
import 'package:googleapis_auth/googleapis_auth.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';
import '../database/database_helper.dart';
import 'crypto_service.dart';

/// Sync status enum
enum SyncStatus { idle, syncing, success, error }

/// Simple sync service: downloads finance.json from Drive, imports into DB.
/// Exports DB back to finance.json, uploads to Drive.
/// NO complex merging — just mirrors EXT's approach.
class SyncService {
  SyncService._();
  static final SyncService instance = SyncService._();

  static const _deviceIdKey = 'sync_device_id';
  static const _lastSyncKey = 'sync_last_modified';
  static const _financeFileName = 'finance.json';
  static const _folderName = 'QLCT';

  String? _deviceId;
  SyncStatus _status = SyncStatus.idle;
  String _lastMessage = '';
  final List<VoidCallback> _listeners = [];

  SyncStatus get status => _status;
  String get lastMessage => _lastMessage;

  void addListener(VoidCallback listener) => _listeners.add(listener);
  void removeListener(VoidCallback listener) => _listeners.remove(listener);
  void _notify() { for (final l in _listeners) l(); }

  Future<String> getDeviceId() async {
    if (_deviceId != null) return _deviceId!;
    final prefs = await SharedPreferences.getInstance();
    _deviceId = prefs.getString(_deviceIdKey);
    if (_deviceId == null) {
      _deviceId = 'android_${const Uuid().v4().substring(0, 8)}';
      await prefs.setString(_deviceIdKey, _deviceId!);
    }
    return _deviceId!;
  }

  Future<String?> getLastSyncTimestamp() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_lastSyncKey);
  }

  Future<void> _setLastSyncTimestamp(String timestamp) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_lastSyncKey, timestamp);
  }

  Future<Map<String, dynamic>?> downloadFinanceJson(GoogleSignInAccount user) async {
    return _downloadFinanceJson(user);
  }

  /// Public raw fetch (no decryption) — used by the PIN unlock flow to obtain
  /// the encrypted envelope for establishFromEnvelope onboarding.
  Future<Map<String, dynamic>?> fetchRemoteRaw(GoogleSignInAccount user) async {
    return _fetchRemoteRaw(user);
  }

  Future<void> setLastSyncTimestamp(String timestamp) async {
    await _setLastSyncTimestamp(timestamp);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // FULL SYNC — 2-way merge by UUID + updatedAt
  // ═══════════════════════════════════════════════════════════════════════════

  Future<String> fullSync(GoogleSignInAccount user) async {
    if (_status == SyncStatus.syncing) return 'Đang đồng bộ...';
    _status = SyncStatus.syncing;
    _lastMessage = 'Đang đồng bộ...';
    _notify();

    try {
      final db = DatabaseHelper.instance;

      // Step 1: Export local data
      debugPrint('[SYNC] Exporting local data...');
      final localData = await db.exportFinanceJson();
      final localRecCount = (localData['records'] as List?)?.length ?? 0;
      debugPrint('[SYNC] Local export: $localRecCount records');

      // Step 2: Download remote finance.json (raw, then decrypt if needed)
      debugPrint('[SYNC] Downloading remote...');
      final crypto = CryptoService.instance;
      final remoteRaw = await _fetchRemoteRaw(user);
      // SAFETY: encrypted remote we can't read yet → abort (never clobber ciphertext).
      if (remoteRaw != null &&
          crypto.isEncryptedEnvelope(remoteRaw) &&
          !crypto.hasKey()) {
        _status = SyncStatus.error;
        _lastMessage = 'Dữ liệu Drive đã mã hóa — cần nhập mã PIN để đồng bộ.';
        _notify();
        return _lastMessage;
      }
      Map<String, dynamic>? remoteData;
      if (remoteRaw != null) {
        remoteData = crypto.isEncryptedEnvelope(remoteRaw)
            ? await crypto.decryptData(remoteRaw)
            : remoteRaw;
      }

      Map<String, dynamic> dataToUpload;

      if (remoteData == null) {
        // No remote — upload local (first sync)
        dataToUpload = localData;
        dataToUpload['deviceId'] = await getDeviceId();
        debugPrint('[SYNC] No remote, uploading local');
      } else {
        // Merge: remote is base, add local-only records
        debugPrint('[SYNC] Merging...');
        dataToUpload = _merge(localData, remoteData);
      }

      // Step 3: (Removed old 50%-safety reimport-remote guard.)
      // The merge is a per-record UNION by id — it can only have fewer records
      // than remote when records were soft-deleted (legitimate tombstones).
      // The old guard discarded freshly-added local records that hadn't been
      // pushed yet ("nhập chi tiêu → đồng bộ → mất"). We trust the union merge.
      final mergedRecordCount = (dataToUpload['records'] as List?)?.length ?? 0;

      // Step 4: Upload merged
      debugPrint('[SYNC] Uploading...');
      final uploaded = await _uploadFinanceJson(user, dataToUpload);
      if (!uploaded) {
        _status = SyncStatus.error;
        _lastMessage = 'Upload thất bại';
        _notify();
        return _lastMessage;
      }

      // Step 5: Import merged into local DB
      debugPrint('[SYNC] Importing merged into DB...');
      await db.importFinanceJson(dataToUpload);

      // Step 6: Save timestamp
      await _setLastSyncTimestamp(
        dataToUpload['lastModified'] as String? ?? DateTime.now().toUtc().toIso8601String()
      );

      _status = SyncStatus.success;
      // Count sync stats
      final localRecordCount = (localData['records'] as List?)?.length ?? 0;
      final remoteRecordCount2 = remoteData != null ? ((remoteData['records'] as List?)?.length ?? 0) : 0;
      final addedToRemote = mergedRecordCount > remoteRecordCount2 ? mergedRecordCount - remoteRecordCount2 : 0;
      final addedFromRemote = mergedRecordCount > localRecordCount ? mergedRecordCount - localRecordCount : 0;
      // Count updated records (local newer than remote)
      int updatedCount = 0;
      if (remoteData != null) {
        final remoteRecords = (remoteData['records'] as List<dynamic>? ?? []).cast<Map<String, dynamic>>();
        final remoteMap = <String, String>{};
        for (final r in remoteRecords) { remoteMap[r['id'] as String? ?? ''] = r['updatedAt'] as String? ?? ''; }
        final localRecords = (localData['records'] as List<dynamic>? ?? []).cast<Map<String, dynamic>>();
        for (final r in localRecords) {
          final id = r['id'] as String? ?? '';
          if (remoteMap.containsKey(id)) {
            final lt = r['updatedAt'] as String? ?? '';
            final rt = remoteMap[id] ?? '';
            if (lt.compareTo(rt) > 0) updatedCount++;
          }
        }
      }
      _lastMessage = '✓ ${addedToRemote > 0 ? "$addedToRemote mới" : ""}${updatedCount > 0 ? " $updatedCount cập nhật" : ""} lên Drive | ${addedFromRemote > 0 ? "$addedFromRemote từ Drive" : "0 từ Drive"} ($mergedRecordCount giao dịch)';
      if (addedToRemote == 0 && updatedCount == 0 && addedFromRemote == 0) {
        _lastMessage = '✓ Đã đồng bộ — không có thay đổi ($mergedRecordCount giao dịch)';
      }
      _notify();
      return _lastMessage;

    } catch (e) {
      debugPrint('[SYNC] Error: $e');
      _status = SyncStatus.error;
      _lastMessage = 'Lỗi: $e';
      _notify();
      return _lastMessage;
    }
  }

  /// Quick push: export local → merge with remote → upload
  Future<void> quickPush(GoogleSignInAccount user) async {
    try {
      final db = DatabaseHelper.instance;
      final localData = await db.exportFinanceJson();
      localData['deviceId'] = await getDeviceId();

      final crypto = CryptoService.instance;
      final remoteRaw = await _fetchRemoteRaw(user);
      // SAFETY: don't overwrite an encrypted Drive envelope with plaintext when locked.
      if (remoteRaw != null &&
          crypto.isEncryptedEnvelope(remoteRaw) &&
          !crypto.hasKey()) {
        debugPrint('[SYNC] Quick push aborted: remote is encrypted and locked');
        return;
      }
      final remoteData = (remoteRaw != null && crypto.isEncryptedEnvelope(remoteRaw))
          ? await crypto.decryptData(remoteRaw)
          : remoteRaw;
      Map<String, dynamic> dataToUpload;

      if (remoteData != null) {
        dataToUpload = _merge(localData, remoteData);
      } else {
        dataToUpload = localData;
      }

      await _uploadFinanceJson(user, dataToUpload);
      // Write the merged result back to the local DB so records pulled in from
      // remote during this push are not lost on the next open (previously
      // quickPush uploaded the merge but never re-imported it locally).
      if (remoteData != null) {
        await db.importFinanceJson(dataToUpload);
      }
      await _setLastSyncTimestamp(DateTime.now().toUtc().toIso8601String());
    } catch (e) {
      debugPrint('[SYNC] Quick push error: $e');
    }
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // MERGE — same as EXT: UUID + updatedAt, remote as base
  // ═══════════════════════════════════════════════════════════════════════════

  Map<String, dynamic> _merge(Map<String, dynamic> local, Map<String, dynamic> remote) {
    // Start from remote (preserves all EXT-specific data)
    final merged = Map<String, dynamic>.from(remote);

    // Merge records by UUID + updatedAt
    final localRecords = (local['records'] as List<dynamic>? ?? [])
        .cast<Map<String, dynamic>>();
    final remoteRecords = (remote['records'] as List<dynamic>? ?? [])
        .cast<Map<String, dynamic>>();

    final localMap = <String, Map<String, dynamic>>{};
    for (final r in localRecords) {
      final id = r['id'] as String?;
      if (id != null) localMap[id] = r;
    }
    final remoteMap = <String, Map<String, dynamic>>{};
    for (final r in remoteRecords) {
      final id = r['id'] as String?;
      if (id != null) remoteMap[id] = r;
    }

    final mergedRecords = <String, Map<String, dynamic>>{};

    // All remote records (remote wins on equal timestamps)
    for (final entry in remoteMap.entries) {
      final id = entry.key;
      if (localMap.containsKey(id)) {
        final localRec = localMap[id]!;
        final remoteRec = entry.value;
        final localModuleId = localRec['moduleId'] as String? ?? '';
        final remoteModuleId = remoteRec['moduleId'] as String? ?? '';

        // MODULE CONFLICT: if moduleId differs, remote (EXT) is source of truth
        if (localModuleId != remoteModuleId && remoteModuleId.isNotEmpty) {
          mergedRecords[id] = remoteRec;
        } else {
          // Same module: use updatedAt to resolve (last-write-wins).
          final lt = _parseTime(localRec['updatedAt']);
          final rt = _parseTime(remoteRec['updatedAt']);
          if (lt.isAfter(rt)) {
            mergedRecords[id] = localRec;
          } else if (rt.isAfter(lt)) {
            mergedRecords[id] = remoteRec;
          } else {
            // Exact tie: tombstone wins so a delete isn't silently undone;
            // otherwise keep local (this device's edit).
            final localDeleted = localRec['isDeleted'] == true;
            final remoteDeleted = remoteRec['isDeleted'] == true;
            if (localDeleted && !remoteDeleted) {
              mergedRecords[id] = localRec;
            } else if (remoteDeleted && !localDeleted) {
              mergedRecords[id] = remoteRec;
            } else {
              mergedRecords[id] = localRec;
            }
          }
        }
      } else {
        mergedRecords[id] = entry.value;
      }
    }

    // Local-only records
    for (final entry in localMap.entries) {
      if (!remoteMap.containsKey(entry.key)) {
        mergedRecords[entry.key] = entry.value;
      }
    }

    merged['records'] = mergedRecords.values.toList();

    // Merge accounts by id + updatedAt
    final localAccounts = (local['accounts'] as List<dynamic>? ?? []).cast<Map<String, dynamic>>();
    final remoteAccounts = (remote['accounts'] as List<dynamic>? ?? []).cast<Map<String, dynamic>>();
    merged['accounts'] = _mergeById(localAccounts, remoteAccounts);

    // Merge modules by id (union) so locally-created modules aren't wiped out
    // by remote. Remote is base; add any local-only module (e.g. "Tiết Kiệm").
    final localModules = (local['modules'] as List<dynamic>? ?? []).cast<Map<String, dynamic>>();
    final remoteModules = (remote['modules'] as List<dynamic>? ?? []).cast<Map<String, dynamic>>();
    if (localModules.isNotEmpty || remoteModules.isNotEmpty) {
      final byId = <String, Map<String, dynamic>>{};
      for (final m in remoteModules) {
        final id = m['id'] as String?;
        if (id != null) byId[id] = m;
      }
      for (final m in localModules) {
        final id = m['id'] as String?;
        if (id != null && !byId.containsKey(id)) byId[id] = m; // add local-only
      }
      merged['modules'] = byId.values.toList();
    }

    merged['lastModified'] = DateTime.now().toUtc().toIso8601String();

    return merged;
  }

  List<Map<String, dynamic>> _mergeById(List<Map<String, dynamic>> local, List<Map<String, dynamic>> remote) {
    final localMap = <String, Map<String, dynamic>>{};
    for (final item in local) { final id = item['id'] as String?; if (id != null) localMap[id] = item; }
    final remoteMap = <String, Map<String, dynamic>>{};
    for (final item in remote) { final id = item['id'] as String?; if (id != null) remoteMap[id] = item; }

    final merged = <String, Map<String, dynamic>>{};
    for (final entry in remoteMap.entries) {
      final id = entry.key;
      if (localMap.containsKey(id)) {
        final lt = _parseTime(localMap[id]!['updatedAt']);
        final rt = _parseTime(entry.value['updatedAt']);
        merged[id] = lt.isAfter(rt) ? localMap[id]! : entry.value;
      } else {
        merged[id] = entry.value;
      }
    }
    for (final entry in localMap.entries) {
      if (!remoteMap.containsKey(entry.key)) merged[entry.key] = entry.value;
    }
    return merged.values.toList();
  }

  DateTime _parseTime(dynamic value) {
    if (value == null) return DateTime(2000);
    try { return DateTime.parse(value.toString()); } catch (_) { return DateTime(2000); }
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // GOOGLE DRIVE — download/upload finance.json
  // ═══════════════════════════════════════════════════════════════════════════

  /// Fetch the remote finance file content WITHOUT decrypting.
  /// Returns the parsed JSON (an encrypted envelope OR plaintext finance data),
  /// or null if there is no remote file. Used to detect whether Drive holds an
  /// encrypted envelope this client cannot read yet — so sync never clobbers
  /// ciphertext with plaintext.
  Future<Map<String, dynamic>?> _fetchRemoteRaw(GoogleSignInAccount user) async {
    final client = await _getAuthClient(user);
    if (client == null) return null;

    try {
      final driveApi = drive.DriveApi(client);
      final folderId = await _findOrCreateFolder(driveApi);

      String query = "name = '$_financeFileName' and trashed = false";
      if (folderId != null) query += " and '$folderId' in parents";

      final fileList = await driveApi.files.list(
        q: query, orderBy: 'modifiedTime desc', pageSize: 1,
        $fields: 'files(id, name, modifiedTime)',
      );

      final files = fileList.files;
      if (files == null || files.isEmpty) return null;

      final fileId = files.first.id!;
      final media = await driveApi.files.get(fileId, downloadOptions: drive.DownloadOptions.fullMedia) as drive.Media;

      final bytes = <int>[];
      await for (final chunk in media.stream) { bytes.addAll(chunk); }
      final jsonStr = utf8.decode(bytes);
      return jsonDecode(jsonStr) as Map<String, dynamic>;
    } catch (e) {
      debugPrint('[SYNC] Download error: $e');
      return null;
    } finally {
      client.close();
    }
  }

  /// True when Drive holds an encrypted envelope this device cannot decrypt yet
  /// (no PIN key loaded). In that state we must NOT push plaintext over it.
  Future<bool> isRemoteLocked(GoogleSignInAccount user) async {
    final raw = await _fetchRemoteRaw(user);
    return raw != null &&
        CryptoService.instance.isEncryptedEnvelope(raw) &&
        !CryptoService.instance.hasKey();
  }

  /// Download finance.json and, if it is an encrypted envelope, decrypt it.
  /// Returns null if there is no remote file OR if it is encrypted but this
  /// device has no key loaded (locked) — callers guard with [isRemoteLocked].
  Future<Map<String, dynamic>?> _downloadFinanceJson(GoogleSignInAccount user) async {
    final raw = await _fetchRemoteRaw(user);
    if (raw == null) return null;
    final crypto = CryptoService.instance;
    if (crypto.isEncryptedEnvelope(raw)) {
      if (!crypto.hasKey()) return null; // locked
      return await crypto.decryptData(raw);
    }
    return raw;
  }

  Future<bool> _uploadFinanceJson(GoogleSignInAccount user, Map<String, dynamic> data) async {
    final client = await _getAuthClient(user);
    if (client == null) return false;

    try {
      final driveApi = drive.DriveApi(client);
      final folderId = await _findOrCreateFolder(driveApi);
      // Encrypt to an envelope when PIN encryption is enabled; else plaintext.
      final crypto = CryptoService.instance;
      final Map<String, dynamic> payload =
          crypto.hasKey() ? await crypto.encryptData(data) : data;
      final content = utf8.encode(jsonEncode(payload));
      final media = drive.Media(Stream.value(content), content.length);

      // Find existing file
      String query = "name = '$_financeFileName' and trashed = false";
      if (folderId != null) query += " and '$folderId' in parents";
      final fileList = await driveApi.files.list(q: query, pageSize: 1, $fields: 'files(id)');

      if (fileList.files != null && fileList.files!.isNotEmpty) {
        await driveApi.files.update(drive.File(), fileList.files!.first.id!, uploadMedia: media);
      } else {
        final fileMetadata = drive.File()
          ..name = _financeFileName
          ..parents = folderId != null ? [folderId] : null;
        await driveApi.files.create(fileMetadata, uploadMedia: media);
      }
      return true;
    } catch (e) {
      debugPrint('[SYNC] Upload error: $e');
      return false;
    } finally {
      client.close();
    }
  }

  Future<http.Client?> _getAuthClient(GoogleSignInAccount user) async {
    try {
      // Force refresh authentication to get fresh token
      final auth = await user.authentication;
      final token = auth.accessToken;
      if (token == null) {
        debugPrint('[SYNC] Auth: no access token');
        return null;
      }
      debugPrint('[SYNC] Auth: got access token (${token.substring(0, 10)}...)');
      final credentials = AccessCredentials(
        AccessToken('Bearer', token, DateTime.now().add(const Duration(hours: 1)).toUtc()),
        null, ['https://www.googleapis.com/auth/drive'],
      );
      return authenticatedClient(http.Client(), credentials);
    } catch (e) {
      debugPrint('[SYNC] Auth error: $e');
      return null;
    }
  }

  Future<String?> _findOrCreateFolder(drive.DriveApi driveApi) async {
    try {
      final folderList = await driveApi.files.list(
        q: "name = '$_folderName' and mimeType = 'application/vnd.google-apps.folder' and trashed = false",
        pageSize: 1, $fields: 'files(id)',
      );
      if (folderList.files != null && folderList.files!.isNotEmpty) {
        return folderList.files!.first.id;
      }
      final folder = drive.File()..name = _folderName..mimeType = 'application/vnd.google-apps.folder';
      final created = await driveApi.files.create(folder);
      return created.id;
    } catch (e) {
      debugPrint('[SYNC] Folder error: $e');
      return null;
    }
  }
}

File: webapp/src/services/sync/syncService.ts
import type { FinanceData } from '@/types';
import { driveService } from '@/services/drive/driveService';
import { indexedDBService } from '@/services/indexeddb/indexedDBService';
import { cryptoService } from '@/services/crypto/cryptoService';

/**
 * Validate that a finance data object has minimum required structure.
 * Only rejects data truly unusable (missing core arrays).
 * Settings can be migrated/defaulted by appStore — NOT a reject reason.
 */
export function validateFinanceData(data: unknown): data is FinanceData {
  if (!data || typeof data !== 'object') return false;
  const d = data as Record<string, unknown>;
  if (!Array.isArray(d.modules)) return false;
  if (!Array.isArray(d.records)) return false;
  if (!Array.isArray(d.accounts)) return false;
  return true;
}

export type SyncStatus = 'idle' | 'syncing' | 'success' | 'conflict' | 'error' | 'locked';

export interface SyncResult {
  status: SyncStatus;
  message: string;
  data?: FinanceData;
}

/**
 * Sync Engine
 * 
 * Strategy: Last-Write-Wins with timestamp comparison.
 * 
 * Flow:
 * 1. On open: Check Drive for newer version → download if newer
 * 2. On edit: Write to IndexedDB immediately → queue Drive upload
 * 3. Conflict: Compare lastModified timestamps
 * 
 * Future enhancement: Field-level merge for true conflict resolution.
 */
class SyncService {
  private syncTimeout: ReturnType<typeof setTimeout> | null = null;
  private readonly DEBOUNCE_MS = 3000; // Wait 3s after last edit before syncing

  /**
   * True when Drive holds an encrypted envelope this client cannot decrypt yet
   * (no PIN key loaded). In that state we must NOT push plaintext over it.
   */
  private async isRemoteLocked(): Promise<boolean> {
    const raw = await driveService.fetchRemoteRaw();
    return !!raw && cryptoService.isEncryptedEnvelope(raw) && !cryptoService.hasKey();
  }

  /**
   * Pull latest data from Google Drive
   * Called on app open or manual sync.
   * Works even when local database is empty (fresh install).
   */
  async pull(): Promise<SyncResult> {
    try {
      const token = driveService.token;
      if (!token) {
        return { status: 'idle', message: 'Not authenticated' };
      }

      const remoteFile = await driveService.findFile();
      if (!remoteFile) {
        // No file on Drive yet
        return { status: 'idle', message: 'No remote file found on Drive' };
      }

      // Encrypted Drive data we can't read yet → require PIN (don't clobber).
      if (await this.isRemoteLocked()) {
        return { status: 'locked', message: 'Dữ liệu Drive đã mã hóa — cần nhập mã PIN.' };
      }

      const localData = await indexedDBService.loadData();
      const localModified = localData?.lastModified
        ? new Date(localData.lastModified).getTime()
        : 0;
      const remoteModified = new Date(remoteFile.modifiedTime).getTime();

      // Download if: no local data (fresh install) OR remote is newer
      if (!localData || remoteModified > localModified) {
        const remoteData = await driveService.downloadFile(remoteFile.id);
        if (remoteData && validateFinanceData(remoteData)) {
          // SAFETY: never blindly replace local with remote — merge by id so
          // freshly-added local records (not yet pushed) are preserved.
          const finalData = localData ? this.mergeData(localData, remoteData).data : remoteData;
          await indexedDBService.saveData(finalData);
          const recordCount = finalData.records?.length ?? 0;
          const moduleCount = finalData.modules?.length ?? 0;
          return {
            status: 'success',
            message: `Đã tải từ Drive: ${moduleCount} modules, ${recordCount} records`,
            data: finalData,
          };
        } else {
          return { status: 'error', message: 'Failed to download file from Drive' };
        }
      }

      return { status: 'success', message: 'Local data is up to date' };
    } catch (error) {
      return {
        status: 'error',
        message: error instanceof Error ? error.message : 'Pull failed',
      };
    }
  }

  /**
   * Push local data to Google Drive
   * SAFETY: refuses to push if local has fewer records than remote (prevents data loss)
   */
  async push(): Promise<SyncResult> {
    try {
      const token = driveService.token;
      if (!token) {
        return { status: 'idle', message: 'Not authenticated' };
      }

      const localData = await indexedDBService.loadData();
      if (!localData) {
        return { status: 'error', message: 'No local data to push' };
      }

      // SAFETY: never overwrite an encrypted Drive envelope with plaintext when locked.
      if (await this.isRemoteLocked()) {
        return { status: 'locked', message: 'Dữ liệu Drive đã mã hóa — cần nhập mã PIN trước khi đồng bộ.' };
      }

      // SAFETY CHECK: don't push if local has significantly fewer records than remote
      const remoteFile = await driveService.findFile();
      if (remoteFile) {
        const remoteData = await driveService.downloadFile(remoteFile.id);
        if (remoteData) {
          const localRecords = localData.records?.length ?? 0;
          const remoteRecords = remoteData.records?.length ?? 0;
          // If remote has records but local is empty or much smaller → REFUSE to push
          if (remoteRecords > 5 && localRecords < remoteRecords * 0.5) {
            return { status: 'error', message: `An toàn: không push vì local (${localRecords} records) ít hơn remote (${remoteRecords} records). Dùng "Đồng bộ ngay" thay vì push.` };
          }
        }
      }

      const fileId = await driveService.uploadFile(localData);
      if (fileId) {
        return { status: 'success', message: 'Pushed to Drive' };
      }

      return { status: 'error', message: 'Upload failed' };
    } catch (error) {
      return {
        status: 'error',
        message: error instanceof Error ? error.message : 'Push failed',
      };
    }
  }

  /**
   * Schedule a push with debouncing.
   * Prevents flooding Drive API during rapid edits.
   */
  schedulePush(): void {
    if (this.syncTimeout) {
      clearTimeout(this.syncTimeout);
    }

    this.syncTimeout = setTimeout(async () => {
      await this.push();
    }, this.DEBOUNCE_MS);
  }

  /**
   * Full sync cycle: pull remote → merge by UUID+updatedAt → push merged → save local
   * If local database is empty (fresh install), does a pull-only operation.
   */
  async fullSync(): Promise<SyncResult> {
    try {
      const token = driveService.token;
      if (!token) {
        return { status: 'idle', message: 'Not authenticated' };
      }

      // SAFETY: never overwrite an encrypted Drive envelope with plaintext when locked.
      if (await this.isRemoteLocked()) {
        return { status: 'locked', message: 'Dữ liệu Drive đã mã hóa — cần nhập mã PIN trước khi đồng bộ.' };
      }

      const localData = await indexedDBService.loadData();

      // Download remote
      const remoteFile = await driveService.findFile();
      let remoteData: FinanceData | null = null;
      if (remoteFile) {
        const downloaded = await driveService.downloadFile(remoteFile.id);
        // Validate remote data has required structure (reject empty/reset payloads)
        if (downloaded && validateFinanceData(downloaded)) {
          remoteData = downloaded;
        } else if (downloaded) {
          console.warn('[SYNC] Remote finance.json invalid (missing settings/modules). Treating as empty.');
        }
      }

      // Case 1: No local data (fresh install)
      if (!localData) {
        if (remoteData) {
          // Fresh install + remote exists → just import remote data
          await indexedDBService.saveData(remoteData);
          const recordCount = remoteData.records?.length ?? 0;
          const moduleCount = remoteData.modules?.length ?? 0;
          return {
            status: 'success',
            message: `Initial sync: ${moduleCount} modules, ${recordCount} records imported`,
            data: remoteData,
          };
        } else {
          // Fresh install + no remote → nothing to do
          return { status: 'idle', message: 'No data available (empty Drive and empty local)' };
        }
      }

      // Case 2: Local data exists
      let mergedData: FinanceData;
      let syncStats = { addedToRemote: 0, addedFromRemote: 0, updated: 0 };
      if (!remoteData) {
        // No remote - just push local
        mergedData = { ...localData, lastModified: new Date().toISOString() };
        syncStats.addedToRemote = localData.records?.length ?? 0;
      } else {
        // Merge records by UUID + updatedAt
        const result = this.mergeData(localData, remoteData);
        mergedData = result.data;
        syncStats = { addedToRemote: result.addedToRemote, addedFromRemote: result.addedFromRemote, updated: result.updated };
      }

      // NOTE: The merge is a per-record UNION by id (see mergeData) — it can
      // never contain FEWER records than remote unless records were soft-deleted
      // (tombstones), which is legitimate. The old "merged < 50% of remote →
      // reimport remote wholesale" guard was DISCARDING freshly-added local
      // records that hadn't been pushed yet (the "add → sync → gone" bug).
      // We trust the union merge and always upload it.

      // Upload merged
      const fileId = await driveService.uploadFile(mergedData);
      if (!fileId) {
        return { status: 'error', message: 'Upload failed' };
      }

      // Save merged locally
      await indexedDBService.saveData(mergedData);

      return { status: 'success', message: `↑${syncStats.addedToRemote} lên Drive | ↓${syncStats.addedFromRemote} từ Drive${syncStats.updated > 0 ? ` | ${syncStats.updated} cập nhật` : ''}`, data: mergedData };
    } catch (error) {
      return {
        status: 'error',
        message: error instanceof Error ? error.message : 'Sync failed',
      };
    }
  }

  /**
   * Merge local and remote FinanceData by UUID + updatedAt
   */
  private mergeData(local: FinanceData, remote: FinanceData): { data: FinanceData; addedToRemote: number; addedFromRemote: number; updated: number } {
    const merged: FinanceData = { ...remote };

    // Merge records by id + updatedAt — track counts
    const localMap = new Map(local.records.map((r) => [r.id, r]));
    const remoteMap = new Map(remote.records.map((r) => [r.id, r]));
    let addedToRemote = 0; // local-only → will be uploaded
    let addedFromRemote = 0; // remote-only → will be downloaded
    let updated = 0; // conflicts resolved

    const mergedRecords = new Map<string, typeof local.records[0]>();

    // All remote records
    for (const [id, remoteRec] of remoteMap) {
      const localRec = localMap.get(id);
      if (localRec) {
        const lt = new Date(localRec.updatedAt || localRec.createdAt || '2000-01-01').getTime();
        const rt = new Date(remoteRec.updatedAt || remoteRec.createdAt || '2000-01-01').getTime();
        if (lt > rt) { mergedRecords.set(id, localRec); updated++; }
        else if (rt > lt) { mergedRecords.set(id, remoteRec); }
        else {
          // Exact updatedAt tie: prefer the deleted side (tombstone wins so a
          // delete isn't silently undone); otherwise keep local (this device's edit).
          if (localRec.isDeleted && !remoteRec.isDeleted) mergedRecords.set(id, localRec);
          else if (remoteRec.isDeleted && !localRec.isDeleted) mergedRecords.set(id, remoteRec);
          else mergedRecords.set(id, localRec);
        }
      } else {
        mergedRecords.set(id, remoteRec);
        addedFromRemote++;
      }
    }

    // Local-only records
    for (const [id, localRec] of localMap) {
      if (!remoteMap.has(id)) {
        mergedRecords.set(id, localRec);
        addedToRemote++;
      }
    }

    merged.records = Array.from(mergedRecords.values());

    // Merge accounts
    merged.accounts = this.mergeArrayById(local.accounts, remote.accounts);

    // Merge modules (keep remote field definitions, merge categories)
    merged.modules = remote.modules.map(remoteMod => {
      const localMod = local.modules.find(m => m.id === remoteMod.id);
      if (localMod && localMod.categories && remoteMod.categories) {
        return {
          ...remoteMod,
          categories: this.mergeArrayById(localMod.categories, remoteMod.categories),
        };
      }
      return remoteMod;
    });

    // Add any local-only modules
    for (const localMod of local.modules) {
      if (!merged.modules.find(m => m.id === localMod.id)) {
        merged.modules.push(localMod);
      }
    }

    merged.lastModified = new Date().toISOString();
    merged.deviceId = local.deviceId;

    return { data: merged, addedToRemote, addedFromRemote, updated };
  }

  /**
   * Merge two arrays of objects by 'id' field, using 'updatedAt' for conflict resolution
   */
  private mergeArrayById<T extends { id: string; updatedAt?: string; createdAt?: string }>(
    local: T[],
    remote: T[]
  ): T[] {
    const localMap = new Map<string, T>();
    for (const item of local) {
      localMap.set(item.id, item);
    }

    const remoteMap = new Map<string, T>();
    for (const item of remote) {
      remoteMap.set(item.id, item);
    }

    const merged = new Map<string, T>();

    // Add all remote items, resolve conflicts
    for (const [id, remoteItem] of remoteMap) {
      const localItem = localMap.get(id);
      if (localItem) {
        const localTime = new Date(localItem.updatedAt || localItem.createdAt || '2000-01-01').getTime();
        const remoteTime = new Date(remoteItem.updatedAt || remoteItem.createdAt || '2000-01-01').getTime();
        merged.set(id, localTime > remoteTime ? localItem : remoteItem);
      } else {
        merged.set(id, remoteItem);
      }
    }

    // Add local-only items
    for (const [id, localItem] of localMap) {
      if (!remoteMap.has(id)) {
        merged.set(id, localItem);
      }
    }

    return Array.from(merged.values());
  }

  /**
   * Export data as JSON file (for backup)
   */
  async exportJSON(): Promise<void> {
    const data = await indexedDBService.loadData();
    if (!data) return;

    const blob = new Blob([JSON.stringify(data, null, 2)], { type: 'application/json' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = `finance_backup_${new Date().toISOString().slice(0, 10)}.json`;
    a.click();
    URL.revokeObjectURL(url);
  }

  /**
   * Import data from JSON file
   */
  async importJSON(file: File): Promise<SyncResult> {
    try {
      const text = await file.text();
      const data = JSON.parse(text) as FinanceData;

      // Basic validation
      if (!data.version || !data.modules || !data.records) {
        return { status: 'error', message: 'Invalid data format' };
      }

      await indexedDBService.saveData(data);
      return { status: 'success', message: 'Import successful', data };
    } catch {
      return { status: 'error', message: 'Failed to parse JSON file' };
    }
  }
}

export const syncService = new SyncService();
